File: packages_py/app_yaml_config/src/app_yaml_config/config_resolver.py

from abc import ABC, abstractmethod
from typing import Dict, Any, List, Optional, Tuple, TypeVar, Generic, Union
import os
import copy
import re
from .core import AppYamlConfig
from .domain import BaseResolveOptions, BaseResult, ResolutionSource

TOptions = TypeVar('TOptions', bound=BaseResolveOptions)
TResult = TypeVar('TResult', bound=BaseResult)

class ConfigResolver(ABC, Generic[TOptions, TResult]):
    """Abstract base class for retrieving configuration with env overwrites and fallbacks."""

    def __init__(self, config: Optional[AppYamlConfig] = None):
        self.config = config or AppYamlConfig.get_instance()

    # ========== Abstract Properties ==========

    @property
    @abstractmethod
    def root_key(self) -> str:
        """Root key in YAML config (e.g., 'providers', 'services', 'storage')."""
        pass

    @property
    @abstractmethod
    def meta_key_pattern(self) -> Dict[str, Any]:
        """
        Meta key pattern for env overwrite detection.
        Example: 
          {'type': 'grouped', 'keys': {'overwrite': '...', 'fallbacks': '...'}}
          {'type': 'per-property', 'regex': re.Pattern}
        """
        pass

    @property
    @abstractmethod
    def not_found_error(self) -> type:
        """Exception class to raise when item is not found."""
        pass

    @abstractmethod
    def build_result(
        self,
        name: str,
        config: Dict[str, Any],
        env_overwrites: List[str],
        resolution_sources: Dict[str, ResolutionSource],
        options: TOptions
    ) -> TResult:
        """Build the final result object."""
        pass

    # ========== Public API ==========

    def get(self, name: str, options: Optional[TOptions] = None) -> TResult:
        """Get a configuration item by name with full resolution."""
        opts = self.get_default_options(options)
        items = self.config.get(self.root_key) or {}
        item_raw = items.get(name)

        if not item_raw:
            raise self.not_found_error(name)

        # Deep copy to avoid mutation
        result = copy.deepcopy(item_raw)

        # Hook for subclass preprocessing (e.g., global merge)
        result = self.pre_process(result, opts)

        # Extract env meta based on pattern
        env_meta = self._extract_env_meta(result)

        # Apply env resolution
        env_overwrites, resolution_sources = self._apply_env_resolution(result, env_meta, opts)

        # Remove meta keys if configured
        if opts.remove_meta_keys:
            self._remove_meta_keys(result)

        return self.build_result(name, result, env_overwrites, resolution_sources, opts)

    def list(self) -> List[str]:
        """List all available item names."""
        items = self.config.get(self.root_key) or {}
        return list(items.keys())

    def has(self, name: str) -> bool:
        """Check if an item exists."""
        return name in self.list()

    # ========== Protected Hooks ==========

    def get_default_options(self, options: Optional[TOptions]) -> TOptions:
        """Override to provide default options."""
        return options or BaseResolveOptions()  # type: ignore

    def pre_process(self, config: Dict[str, Any], options: TOptions) -> Dict[str, Any]:
        """Override to modify config before env resolution (e.g. global merge)."""
        return config

    # ========== Shared Implementation ==========

    def _try_env_vars(self, env_vars: Union[str, List[str]]) -> Tuple[Optional[str], Optional[str]]:
        """Try environment variables in order, return first found value and var name."""
        vars_list = env_vars if isinstance(env_vars, list) else [env_vars]
        
        for env_var in vars_list:
            if not isinstance(env_var, str):
                continue
            value = os.environ.get(env_var)
            if value is not None:
                return value, env_var
                
        return None, None

    def _extract_env_meta(self, config: Dict[str, Any]) -> Dict[str, Dict[str, Any]]:
        """Extract environment variable metadata based on the pattern."""
        pattern = self.meta_key_pattern
        result: Dict[str, Dict[str, Any]] = {}

        if pattern.get('type') == 'single':
            key = pattern['key']
            overwrites = config.get(key, {})

            if isinstance(overwrites, dict):
                for prop, env_spec in overwrites.items():
                    if prop not in result:
                        result[prop] = {}
                    result[prop]['primary'] = env_spec

        elif pattern.get('type') == 'grouped':
            # Pattern: overwrite_from_env / fallbacks_from_env
            overwrites = config.get(pattern['keys']['overwrite'], {})
            
            if isinstance(overwrites, dict):
                for prop, env_spec in overwrites.items():
                    if prop not in result:
                        result[prop] = {}
                    result[prop]['primary'] = env_spec

        elif pattern.get('type') == 'per-property':
            # Pattern: env_{prop}_key / env_{prop}_key_fallbacks
            regex = pattern['regex']
            for key, value in config.items():
                match = regex.match(key)
                if match:
                    base_prop = match.group(1)
                    # Check if suffix implies fallback
                    # Regex must be: ^env_(.+)_key(_fallbacks)?$
                    # Group 1 = base_prop
                    # Group 2 = _fallbacks (optional)
                    is_fallback = match.group(2) == '_fallbacks'

                    if base_prop:
                        if base_prop not in result:
                            result[base_prop] = {}

                        if is_fallback:
                            result[base_prop]['fallbacks'] = value
                        else:
                            result[base_prop]['primary'] = value

        return result

    def _apply_env_resolution(
        self,
        result: Dict[str, Any],
        env_meta: Dict[str, Dict[str, Any]],
        options: TOptions
    ) -> Tuple[List[str], Dict[str, ResolutionSource]]:
        """Apply environment variable resolution based on extracted metadata."""
        env_overwrites: List[str] = []
        resolution_sources: Dict[str, ResolutionSource] = {}

        for prop, meta in env_meta.items():
            # Only process if property is explicitly null/None in the config
            # Values set in YAML take precedence over ENV
            if result.get(prop) is not None:
                continue

            # Step 1: Try primary overwrite
            if options.apply_env_overwrites and 'primary' in meta:
                value, matched_var = self._try_env_vars(meta['primary'])
                if value is not None:
                    result[prop] = value
                    env_overwrites.append(prop)
                    resolution_sources[prop] = ResolutionSource(source='env', env_var=matched_var)
                    continue

        return env_overwrites, resolution_sources

    def _remove_meta_keys(self, result: Dict[str, Any]) -> None:
        """Remove metadata keys from the configuration result."""
        pattern = self.meta_key_pattern

        if pattern.get('type') == 'single':
            result.pop(pattern['key'], None)
        elif pattern.get('type') == 'grouped':
            result.pop(pattern['keys']['overwrite'], None)
            result.pop(pattern['keys']['fallbacks'], None)
        elif pattern.get('type') == 'per-property':
            regex = pattern['regex']
            # Find matching keys first to avoid runtime dict size change error
            keys_to_remove = [k for k in result.keys() if regex.match(k)]
            for key in keys_to_remove:
                del result[key]

File: packages_mjs/app-yaml-config/src/config_resolver.ts

import { AppYamlConfig } from './core';
import { BaseResolveOptions, BaseResult, ResolutionSource } from './domain';

// Type describing the meta key pattern
export type MetaKeyPattern =
    | { type: 'grouped'; keys: { overwrite: string; fallbacks: string } }
    | { type: 'single'; key: string }
    | { type: 'per-property'; regex: RegExp };

export abstract class ConfigResolver<TOptions extends BaseResolveOptions, TResult extends BaseResult> {
    protected config: AppYamlConfig;

    constructor(config?: AppYamlConfig) {
        this.config = config || AppYamlConfig.getInstance();
    }

    // ========== Abstract Properties ==========

    protected abstract get rootKey(): string;

    protected abstract get metaKeyPattern(): MetaKeyPattern;

    protected abstract createNotFoundError(name: string): Error;

    protected abstract buildResult(
        name: string,
        config: Record<string, any>,
        envOverwrites: string[],
        resolutionSources: Record<string, ResolutionSource>,
        options: TOptions
    ): TResult;

    // ========== Public API ==========

    public get(name: string, options?: TOptions): TResult {
        const opts = this.getDefaultOptions(options);
        const items = this.config.get(this.rootKey) || {};
        const itemRaw = items[name];

        if (!itemRaw) {
            throw this.createNotFoundError(name);
        }

        // Deep copy (naive JSON approach sufficient for simple config, or explicit copy)
        // Using JSON for now as it handles primitives well, assuming JSON-safe config
        let result = JSON.parse(JSON.stringify(itemRaw));

        // Hook for subclass preprocessing
        result = this.preProcess(result, opts);

        // Extract env meta
        const envMeta = this.extractEnvMeta(result);

        // Apply env resolution
        const { envOverwrites, resolutionSources } = this.applyEnvResolution(result, envMeta, opts);

        // Remove meta keys if configured
        if (opts.removeMetaKeys !== false) {
            this.removeMetaKeys(result);
        }

        return this.buildResult(name, result, envOverwrites, resolutionSources, opts);
    }

    public list(): string[] {
        const items = this.config.get(this.rootKey) || {};
        return Object.keys(items);
    }

    public has(name: string): boolean {
        return this.list().includes(name);
    }

    // ========== Protected Hooks ==========

    protected getDefaultOptions(options?: TOptions): TOptions {
        return (options || {}) as TOptions;
    }

    protected preProcess(config: Record<string, any>, options: TOptions): Record<string, any> {
        return config;
    }

    // ========== Shared Implementation ==========

    protected tryEnvVars(envVars: string | string[]): { value: string | null; envVar: string | null } {
        const varsList = Array.isArray(envVars) ? envVars : [envVars];

        for (const envVar of varsList) {
            const val = process.env[envVar];
            if (val !== undefined && val !== null) {
                return { value: val, envVar };
            }
        }
        return { value: null, envVar: null };
    }

    protected extractEnvMeta(config: Record<string, any>): Record<string, { primary?: string | string[]; fallbacks?: string | string[] }> {
        const pattern = this.metaKeyPattern;
        const result: Record<string, { primary?: any; fallbacks?: any }> = {};

        if (pattern.type === 'single') {
            const overwrites = config[pattern.key] || {};
            for (const [prop, envSpec] of Object.entries(overwrites)) {
                if (!result[prop]) result[prop] = {};
                result[prop].primary = envSpec;
            }
        } else if (pattern.type === 'grouped') {
            const overwrites = config[pattern.keys.overwrite] || {};
            const fallbacks = config[pattern.keys.fallbacks] || {};

            for (const [prop, envSpec] of Object.entries(overwrites)) {
                if (!result[prop]) result[prop] = {};
                result[prop].primary = envSpec;
            }

            for (const [prop, envSpec] of Object.entries(fallbacks)) {
                if (!result[prop]) result[prop] = {};
                result[prop].fallbacks = envSpec;
            }

        } else if (pattern.type === 'per-property') {
            const regex = pattern.regex;
            for (const [key, value] of Object.entries(config)) {
                const match = regex.exec(key);
                if (match) {
                    const baseProp = match[1];
                    const isFallback = match[2] === '_fallbacks';

                    if (baseProp) {
                        if (!result[baseProp]) result[baseProp] = {};
                        if (isFallback) {
                            result[baseProp].fallbacks = value;
                        } else {
                            result[baseProp].primary = value;
                        }
                    }
                }
            }
        }

        return result;
    }

    protected applyEnvResolution(
        result: Record<string, any>,
        envMeta: Record<string, { primary?: any; fallbacks?: any }>,
        options: TOptions
    ) {
        const envOverwrites: string[] = [];
        const resolutionSources: Record<string, ResolutionSource> = {};

        const applyOverwrites = options.applyEnvOverwrites !== false;

        for (const [prop, meta] of Object.entries(envMeta)) {
            // Only process if property is explicitly null/undefined/empty? 
            // Requirement: "if current value is null/None"
            if (result[prop] !== null && result[prop] !== undefined) {
                continue;
            }

            // Step 1: Try primary overwrite
            if (applyOverwrites && meta.primary) {
                const { value, envVar } = this.tryEnvVars(meta.primary);
                if (value !== null) {
                    result[prop] = value;
                    envOverwrites.push(prop);
                    resolutionSources[prop] = { source: 'env', envVar };
                    continue;
                }
            }
        }

        return { envOverwrites, resolutionSources };
    }

    protected removeMetaKeys(result: Record<string, any>): void {
        const pattern = this.metaKeyPattern;

        if (pattern.type === 'single') {
            delete result[pattern.key];
        } else if (pattern.type === 'grouped') {
            delete result[pattern.keys.overwrite];
            delete result[pattern.keys.fallbacks];
        } else if (pattern.type === 'per-property') {
            const regex = pattern.regex;
            const keysToRemove = Object.keys(result).filter(k => regex.test(k));
            keysToRemove.forEach(k => delete result[k]);
        }
    }
}
